parachain-upgrade/src/lib.rs

```diff
@@ -0,0 +1,588 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! Enable parachain validation function upgrades.
+//!
+//! Allow a user to determine when a parachain validation function upgrade
+//! is legal, and perform the upgrade, triggering runtime events
+//! for both storing and applying the new validation function.
+//!
+//! Depends on no external pallets or traits.
+//!
+//! This pallet depends on certain environmental conditions provided by
+//! Cumulus. It will not work outside a Cumulus parachain.
+
+use cumulus_primitives::validation_function_params::{
+	ValidationFunctionParams,
+	NEW_VALIDATION_CODE, VALIDATION_FUNCTION_PARAMS, INHERENT_IDENTIFIER,
+};
+use system::ensure_none;
+use frame_support::{
+	decl_error, decl_event, decl_module, decl_storage, ensure, storage, traits::Get,
+	weights::{SimpleDispatchInfo, Weight, MINIMUM_WEIGHT},
+};
+use parachain::primitives::RelayChainBlockNumber;
+use sp_core::storage::well_known_keys;
+use sp_inherents::{ProvideInherent, InherentData, InherentIdentifier};
+use sp_version::RuntimeVersion;
+
+pub type ValidationFunction = Vec<u8>;
+type System<T> = system::Module<T>;
+
+/// The pallet's configuration trait.
+pub trait Trait: system::Trait {
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+
+	/// Get the chain's current version.
+	type Version: Get<RuntimeVersion>;
+}
+
+// This pallet's storage items.
+decl_storage! {
+	trait Store for Module<T: Trait> as ParachainUpgrade {
+		// we need to store the new validation function for the span between
+		// setting it and applying it.
+		PendingValidationFunction get(fn new_validation_function): (RelayChainBlockNumber, ValidationFunction);
+
+		// we keep the validation function parameters here because we have to use
+		// storage as a transport layer between the inherents and the module
+		VFPs get(fn vfps): ValidationFunctionParams;
+
+		/// Were the VFPs updated this block?
+		DidUpdateVFPs: bool;
+	}
+}
+
+// The pallet's dispatchable functions.
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		// Initializing events
+		// this is needed only if you are using events in your pallet
+		fn deposit_event() = default;
+
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			if let Ok((apply_block, validation_function)) = PendingValidationFunction::try_get() {
+				let relay_chain_height = Self::validation_function_params().relay_chain_height;
+				if relay_chain_height >= apply_block {
```

Comment:
@rphmeier When we later have delayed `PoVBlock` inclusion, the `ValidationParams` given to `validate_block` will always be build based on the block, the `PoVBlock` was build for? Everything else wouldn't make any sense?

Reply:
The `ValidationParams` should always correspond to the state of the `relay_parent` in the `AbridgedCandidateReceipt`, so when we have what I've been calling "contextual execution", where we can include an `AbridgedCandidateReceipt` from the context of a earlier `relay_parent`, the `ValidationParams` will still match that `relay_parent`.

If that's what you mean by delayed `PoVBlock` inclusion, then yes.

Reply:
Yep:) "contextual execution" sounds much better ;)